src/main/kotlin/org/rust/ide/refactoring/move/RsMoveFilesOrDirectoriesProcessor.kt

```diff
@@ -0,0 +1,304 @@
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.psi.util.parentOfType
+import com.intellij.refactoring.RefactoringBundle
+import com.intellij.refactoring.move.MoveCallback
+import com.intellij.refactoring.move.MoveMultipleElementsViewDescriptor
+import com.intellij.refactoring.move.moveFilesOrDirectories.MoveFilesOrDirectoriesProcessor
+import com.intellij.usageView.UsageInfo
+import com.intellij.usageView.UsageViewDescriptor
+import com.intellij.util.IncorrectOperationException
+import com.intellij.util.containers.MultiMap
+import org.rust.ide.inspections.import.insertUseItem
+import org.rust.ide.inspections.import.lastElement
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+
+/**
+ * Move refactoring currently supports moving single file without submodules.
+ * It consists of these steps:
+ * - Check visibility conflicts (for all references their target should remain accessible after move)
+ * - Update `pub(in path)` visibility modifiers in moved file if necessary
+ * - Move mod-declaration to new parent mod
+ * - Update necessary imports in other files
+ * - Update necessary paths in other files (some usages could still remain invalid because of glob-imports)
+ *     We replace path with absolute if there are few usages with same path in file, otherwise add new import
+ * - Update relative paths in moved file
+ */
+class RsMoveFilesOrDirectoriesProcessor(
+    private val project: Project,
+    private val elementsToMove: Array<PsiElement>,
+    private val newParent: PsiDirectory,
+    private val newParentMod: RsMod,
+    moveCallback: MoveCallback?,
+    doneCallback: Runnable
+) : MoveFilesOrDirectoriesProcessor(
+    project,
+    elementsToMove,
+    newParent,
+    true,
+    true,
+    true,
+    moveCallback,
+    doneCallback
+) {
+
+    private val psiFactory = RsPsiFactory(project)
+    private val movedFile: RsFile = elementsToMove[0] as RsFile
+
+    // keys --- `RsPath`s inside movedFile
+    // outsideReferencesMap[path] --- target element for path reference
+    private lateinit var outsideReferencesMap: Map<RsPath, RsElement>
+    private lateinit var conflictsDetector: RsMoveFilesOrDirectoriesConflictsDetector
+
+    override fun doRun() {
+        checkMove()
+        super.doRun()
+    }
+
+    private fun checkMove() {
+        // TODO: support move multiply files
+        check(elementsToMove.size == 1)
+
+        check(newParentMod.crateRoot == movedFile.crateRoot)
+        movedFile.modName?.let {
+            if (newParentMod.getChildModule(it) != null) {
+                throw IncorrectOperationException("Cannot move. Mod with same crate relative path already exists")
+            }
+        }
+    }
+
+    override fun preprocessUsages(refUsages: Ref<Array<UsageInfo>>): Boolean {
+        val usages = refUsages.get()
+        val conflicts = MultiMap<PsiElement, String>()
+
+        val success = ProgressManager.getInstance().runProcessWithProgressSynchronously(
```

Comment:
Looks like it is not necessary to call `checkCanceled` manually (I put breakpoint at `checkCanceled` and it was called from platform code). From [General Threading Rules](https://www.jetbrains.org/intellij/sdk/docs/basics/architectural_overview/general_threading_rules.html#background-processes-and-processcanceledexception):

> PSI internals have a lot of checkCanceled() calls inside

---

Regarding `project.computeWithCancelableProgress`, I want to add new method `project.runWithCancelableProgress` (which will return `Boolean` like currently used `runProcessWithProgressSynchronously` method), what do you think?

Reply:
You're right, calling `checkCanceled` is usually not necessary in case of lots PSI stuff in the background process. However it's better to check that `Cancel` button works in your particular case.

> I want to add new method project.runWithCancelableProgress

OK, I think that's fine